### DISCUSSION TITLE
Working towards freeRouting as a service (FAAS)

### DISCUSSION THREAD

**Discussion by retiredfeline** [freerouting/freerouting#195] [Ideas]
As discussed in the issue containing this comment: https://github.com/freerouting/freerouting/issues/151#issuecomment-1547372702 I'm willing to look into wrapping a network server around freeRouting so that DSNs can be submitted over the network, and SES and log files returned.

To do this freeRouting has to be able to run without any GUI at all, as the execution environment will not have a GUI. Is this possible?

**Comment by retiredfeline** [comment]
So I started looking at what changes are needed in the Java code to run headless. I found this guide:

https://www.oracle.com/technical-resources/articles/javase/headless.html

So I tried:

java -Djava.awt.headless=true -jar freerouting_jar -mt 1 -dl -de foo.dsn -do foo.ses

and it threw an exception:

Exception: java.awt.HeadlessException thrown from the UncaughtExceptionHandler in thread "main"

So a lot needs to be done before I can run freeRouting on a headless server.

**Comment by andrasfuchs** [comment]
Great start!

I'll check out that exception, hoping to learn from it. If you don't use the `-dl` parameter you might find the exact exception in the log file.

I think we would need to separate all the non-GUI parts into a library and build two executables that use that library. 
One executable would be similar to the one we have now (with GUI), and the other, new one would be headless CLI.

**Comment by retiredfeline** [comment]
Not an awful lot in freerouting.log.

```
2023-05-16 03:49:34.871 [main] INFO  Freerouting v1.7.0 (build-date: 2023-02-08)
2023-05-16 03:49:34.873 [main] DEBUG  Version: 1.7.0,2023-02-08
2023-05-16 03:49:34.873 [main] DEBUG  Command line arguments: '-mt 1 -de alarmboard.dsn -do alarmboard.ses'
2023-05-16 03:49:34.874 [main] DEBUG  Architecture: Linux,amd64,5.14.21-150400.24.63-default
2023-05-16 03:49:34.874 [main] DEBUG  Java: 17.0.7,N/A
2023-05-16 03:49:34.875 [main] DEBUG  System Language: en,en_AU
2023-05-16 03:49:34.877 [main] DEBUG  Hardware: 12 CPU cores,8008 MB RAM
2023-05-16 03:49:34.879 [main] DEBUG  UTC Time: 2023-05-15T17:49:34.877695588Z
2023-05-16 03:49:34.880 [main] DEBUG  GUI Language: en_AU
2023-05-16 03:49:34.882 [main] INFO  Opening 'alarmboard.dsn'...
```

**Comment by retiredfeline** [comment]
Here's a fork which is said to run headless from the CLI. Might be worth seeing what they did and whether freerouting can adopt some of the modifications.

http://repo.hu/projects/freerouting_cli/

**Comment by retiredfeline** [comment]
I haven't looked into the code to see if freeRouting is using AWT, or the newer Swing toolkit. Probably the latter since the widgets don't look native. So I did a quick search for a Swing implementation that doesn't need a GUI and found this:

https://github.com/CaciocavalloSilano/caciocavallo

which renders to a virtual screen. Haven't looked to see what it does when input interaction is requested.
